src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py

```diff
@@ -136,10 +140,13 @@ def to_sql(value: Any, datatype: DataType, from_values_statement: bool = False)
         return f"'{binascii.hexlify(value).decode()}' :: BINARY"
 
     if isinstance(value, (list, tuple, array)) and isinstance(datatype, ArrayType):
-        return f"PARSE_JSON({str_to_sql(json.dumps(value))}) :: ARRAY"
+        return f"PARSE_JSON({str_to_sql(json.dumps(value, cls=PythonObjJSONEncoder))}) :: ARRAY"
```

Comment:
I'm not sure whether we need it, because `list, tuple, array, dict` are all json serializable.

Reply:
they aren't if they contain not JSON serializable objects (e.g., date). This is a fix for these scenarios. Shall I add it to bugfix in Changelog?

Reply:
It seems PythonObjJSONEncoder couldn't handle the case that if the container has not JSON serializable objects.

Reply:
Yeah maybe you can mention it in changelog, or put out another PR

Reply:
mentioned in changelog.